content/articles/2018-10-23-dplyr-0-8-0-release-candidate.Rmd

```diff
@@ -0,0 +1,176 @@
+
+```{r}
+df <- tibble(
+  f1 = factor(c("a", "a", "a", "b", "b"), levels = c("a", "b", "c")), 
+  f2 = factor(c("d", "e", "d", "e", "f"), levels = c("d", "e", "f")), 
+  x  = c(1, 1, 1, 2, 2), 
+  y  = 1:5
+)
+```
+
+If we group by the factors `f1` and `f2`, we get 9 groups
+
+```
+# 9 groups : 3 x 3
+df %>% 
+  group_by(f1, f2) %>% 
+  group_data()
+```
+
+When factors and non factors are involved in the grouping, the number of 
+groups depends on the order. At each level of grouping, factors are always expanded
+to one group per level, but non factors only create groups based on observed data. 
+
+```
+# 3 groups
+df %>% 
+  group_by(f1, x) %>% 
+  group_data()
+```
+
+In this example, we group by `f1` then `x`. At the first layer, grouping on `f1` creates
+two groups. Each of these grouops is then subdivided based on the values of the second 
+variable `x`. Since `x` is always 1 when `f1` is `a` the group is not 
+further divided. 
+
+The last group, associated with the level `c` of the factor `f1` is empty, and 
+consequently has no values for the vector `x`. In that case, `group_by()` uses 
+`NA`. 
+
+```
+# 6 groups
+df %>% 
+  group_by(x, f1) %>% 
+  group_data()
+```
+
+When we group by `x` then `f1` we initially split the data according to `x` which 
+gives 2 groups. Each of these two groups is then further divided in 3 groups, 
+i.e. one for each level of `f1`. 
+
+# Performance
+
+TODO: 
+ - unwind protect
+
+# Less surprising hybrid evaluation
+
+Hybrid evaluation has been entirely redesigned to be more predictable 
+and less surprising. 
+
+When we evaluate an expression in `mutate()` or `summarise()`, the hybrid system
+first assert that it can entirely evaluate the expression independently with native code, 
+if it can it does. 
+
+```{r}
+iris %>% 
+  group_by(Species) %>% 
+  summarise(Petal.Length = mean(Petal.Length))
+```
+
+The expression `mean(Petal.Length)` is handled by native code because there is an hybrid 
+internal version of `mean`, and its first argument is a column of the data, and the type of 
+that column can be treated by the internal function. 
+
+If the expression can not be evaluated completely, it gets evaluated using R standard 
+evaluation rules in a special environment that makes the columns available (the data mask). 
+
+```{r}
+iris %>% 
+  group_by(Species) %>% 
+  summarise(Petal.Length = sum(Petal.Length) / n())
+```
+
+The expression `sum(Petal.Length) / n()` is not recognized, so evaluation goes through the 
+standard path. For each group, the subset of `Petal.Length` is extracted, then the R function `sum` 
+is called on it and this is divided by the result of calling the `n()` function from `dplyr`. 
+
+Previous versions of `dplyr` attempted to partially evaluate `sum(Petal.Length) / n()` with the 
+hybrid evaluator, because `sum(Petal.Length)` and `n()` can both independently be handled. This was 
+making it hard to obey R semantics because R is too flexible. 
+
+# nest_by, nest_join
+
+TODO
```

Comment:
Mention that `nest_join()` is fundamental in the sense that you can generate many of the other joins from it.